src/rnnlm/arpa-sampling.h

```diff
@@ -0,0 +1,161 @@
+// arpa_sampling.h
+
+// Copyright     2016  Ke Li
+
+// See ../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ARPA_SAMPLING_H_
+#define ARPA_SAMPLING_H_
+
+#include <sys/time.h>
+#include <unistd.h>
+#include "lm/arpa-file-parser.h"
+#include "fst/fstlib.h"
+
+#include <stdlib.h>
+#include <math.h>
+#include <algorithm>
+#include <map>
+#include <string>
+#include <sstream>
+#include <fstream>
+#include <vector>
+#include <iostream>
+
+namespace kaldi {
+
+typedef int32_t int32;
+
+/// A hashing function-object for vectors of ints.
+struct IntVectorHasher {  // hashing function for vector<Int>.
+  size_t operator()(const std::vector<int32> &x) const {
+    size_t ans = 0;
+    typename std::vector<int32>::const_iterator iter = x.begin(), end = x.end();
+    for (; iter != end; ++iter) {
+      ans *= kPrime;
+      ans += *iter;
+    }
+    return ans;
+  }
+ private:
+  static const int kPrime = 7853;
+};
+
+// Predefine some symbol values, because any integer is as good than any other.
+enum {
+  kEps = 0,
+  // kDisambig,
+  kBos, kEos, kUnk
+};
+
+typedef std::vector<int32> HistType;
+typedef unordered_map<int32, std::pair<BaseFloat, BaseFloat> > WordToProbsMap; 
+typedef unordered_map<HistType, WordToProbsMap, IntVectorHasher> NgramType;
+typedef unordered_map<HistType, BaseFloat, IntVectorHasher> HistWeightsType;
+
+class ArpaSampling : public ArpaFileParser {
```

Comment:
What I mean is you should not write a testing function as a member function of a class.